### PR TITLE
packages/core: add unimplemented auth api definitions

### PR DESCRIPTION
**Pull request by Rugvip**
Figured this setup is a good middle ground between providing a unified auth API towards many different auth providers but still allowing the implementation to differ a bit.

### CODE REVIEW

File: packages/core/src/api/apis/definitions/auth.ts [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+/**
+ * Provides authentication towards Github APIs.
+ *
+ * See https://developer.github.com/apps/building-oauth-apps/understanding-scopes-for-oauth-apps/
+ * for a full list of supported scopes.
+ */
+export const githubAuthApiRef = new ApiRef<OAuthApi>({
+  id: 'core.auth.github',
+  description: 'Provides authentication towards Github APIs',
+});
```
Comment (freben) [inline packages/core/src/api/apis/definitions/auth.ts:149]: How will this work of you want to support both github and an internal GHE?

File: packages/core/src/api/apis/definitions/auth.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+import { ApiRef } from '../ApiRef';
+
+/**
+ * An array of scopes, or a scope string formatted according to the
+ * auth provider, which is typically a space separated list.
+ *
+ * See the documentation for each auth provider for the list of scopes
+ * supported by each provider.
+ */
+export type OAuthScopeLike = string | string[];
```
[the remark this answers is not in the document]
Comment (Rugvip) [reply packages/core/src/api/apis/definitions/auth.ts:26]: Yeah idd, simpler

File: packages/core/src/api/apis/definitions/auth.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+/**
+ * This file contains declarations for common interfaces of auth-related APIs.
+ * The declarations should be used to signal which type of authentication and
+ * authorization methods each separate auth provider supports.
+ *
+ * For example, a Google OAuth provider that supports OAuth 2 and OpenID Connect,
+ * would be declared as follows:
+ *
+ * const googleAuthApiRef = new ApiRef<OAuthApi & OpenIDConnectApi>({ ... })
+ */
```
[the remark this answers is not in the document]
Comment (Rugvip) [reply packages/core/src/api/apis/definitions/auth.ts:37]: Yep, didn't actually intend for the api refs to go in this same file from the beginning though

File: packages/core/src/api/apis/definitions/auth.ts (hunk 1 above)
Comment (Rugvip) [reply packages/core/src/api/apis/definitions/auth.ts:149]: Current idea is that plugins depend on this ref, and the app can choose whether GHE or GitHub is the default one, as in the one registered as a part of app. Then we could either allow per-plugin overrides of the implementation, and/or the plugins that want to use the other provider will have to do that explicitly through a different api ref.

File: packages/core/src/api/apis/definitions/auth.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+export type AccessTokenOptions = {
+  // If this is set to true, the user will not be prompted to log in,
+  // and an empty access token will be returned if there is no existing session.
+  //
+  // This can be used to perform a check whether the user is logged in with a set of scopes,
+  // or if you don't want to force a user to be logged in, but provide functionality if they already are.
```
[the remark this answers is not in the document]
Comment (Rugvip) [reply packages/core/src/api/apis/definitions/auth.ts:44]: yep, should've been jsdoc comments too

**Review by marcuseide — APPROVED** [review]

**Pull request merged**